### PR TITLE
A working API to automate the creation of questions

### PR DESCRIPTION
**Pull request by ombhardwajj**
# What
A working API to automate the creation of questions for the all the tables of "subject types" and "Address" has been successfully built.

# How does it work?
1) User sends a POST req at /api/metabase/create-questions
2) It calls "createQuestionsForSubjectTypes" method
3) This method first waits for the Database Sync to complete in metabase and then begins with the main task.
4) First it searches and stores all the names of the "subject types" in a List (since the tables would have the same name).
5) Then it executes the main POST req with payload to metabase which would create the questions/cards

### CODE REVIEW

**Review by vinayvenu — CHANGES_REQUESTED** [review]
DatabaseRepository - Create proper objects for what you return, not JsonNode. 

DatabaseService - Don't return Strings if there are specific known values. Move them into enum values

DatabaseService (createQuestionForTable) - Push towards making smaller objects that describe your data instead of directly using ObjectNodes. 

DatabaseService - Eventually, we will plug this in when there are changes in programs, subject types etc. Which means there needs to be methods to make changes for specific subject types

StringUtils.java - There is an S.java that is meant for String utilities

getDatabaseDetails is being called multiple times. This is a network call, and can potentially get slow. Try to optimize

Not sure of the purpose of the db migrations

application.properties changes need to be reverted

File: avni-server-api/src/main/resources/application.properties [hunk 1]
```diff
@@ -69,15 +69,15 @@ spring.jackson.serialization.fail-on-empty-beans=false
...
+debug=true
 avni.defaultUserName=${OPENCHS_USER_NAME:admin}
 spring.servlet.multipart.max-file-size=10028KB
 spring.servlet.multipart.max-request-size=10028KB
 
 # Network
 server.tomcat.protocol-header=x-forwarded-proto
 spring.security.require-ssl=true
-server.port=${OPENCHS_SERVER_PORT:8021}
+server.port=${OPENCHS_SERVER_PORT:8080}
```
Comment (himeshr) [inline avni-server-api/src/main/resources/application.properties:80]: Revert the changes made to port and debug flag.

Comment (ombhardwajj) [reply avni-server-api/src/main/resources/application.properties]: Had to change the port in my local system as 8021 was already occupied. 
Apologies for pushing this change!

**Comment by ombhardwajj** [comment]
@vinayvenu 
I've fixed the code according to the comments you made.
Regarding db migrations , basically their were some merge conflicts becuase my forked repo was not synced and this has been a result of that. I can remove those files anytime!!

File: avni-server-api/src/main/java/org/avni/server/domain/metabase/TableType.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
+package org.avni.server.domain.metabase;
+
+public enum TableType {
+    INDIVIDUAL("Individual"),
+    HOUSEHOLD("Household"),
```
Comment (himeshr) [inline avni-server-api/src/main/java/org/avni/server/domain/metabase/TableType.java:5]: Where is table for EncounterType

File: avni-server-api/src/main/java/org/avni/server/web/MetabaseController.java [hunk 3]
```diff
@@ -21,5 +26,18 @@ public MetabaseController(MetabaseService metabaseService, UserService userServi
     public void setupMetabase() {
         accessControlService.checkPrivilege(PrivilegeType.EditOrganisationConfiguration);
         metabaseService.setupMetabase();
-}
+    }
+
+    @PostMapping("/create-questions")
+    public void createQuestions() throws Exception{
+        databaseService.createQuestionsForSubjectTypes();
+        databaseService.createQuestionsForProgramsAndEncounters();
```
Comment (himeshr) [inline avni-server-api/src/main/java/org/avni/server/web/MetabaseController.java:34]: databaseService call to createQuestionsForSubjectGeneralEncounterTypes() is missing..

File: avni-server-api/src/main/java/org/avni/server/domain/metabase/TableType.java (hunk 2 above)
Comment (ombhardwajj) [reply avni-server-api/src/main/java/org/avni/server/domain/metabase/TableType.java:5]: We have "ProgramEncounter" isn't it? 🤔

**Comment by himeshr** [comment]
@ombhardwajj Pretty happy with the overall design and functionality of the Automated Question Creation service on Metabase.

Have specified list of code review comments from our code walkthrough session earlier:

- Rename createQuestionForIndividualTable to non Individual something, DatabaseRepository
- For Enums and other constants created to represent Metabase entityTypes, provide documentation links.
- Metbase client jar is available, for Java8 avni version and If easy, only then replace..
- Delete MetabaseJoin.java
- Use ConditionType.equal in MetabaseQueryBuilder => conditionArray
- Replace "base-type" with Enums in MetabaseQueryBuilder
- Create constants wherever demmed fit MetabaseQueryBuilder
- Link doc to MetabaseRequestBody
- See if TableDetails is needed in fetchingTableDetails calls from Metabase
- TableType.Encounter and its questions creation
- Move AddressQuestionCreationService logic to DatabaseService
- Replace getSubjectTypeNames usage of int literals with constants depictiing columnsNames
- Create common method called by createQuestionsForSubjectTypes() and createQuestionsForPrograms(), createQuestionsForProgramsEncounters() and createQuestionsForEncounters() with List<String> entityNames
- createQuestionForTable( schema => !public is not used)

**Comment by himeshr** [comment]
@ombhardwajj another thought that occurred to me regarding this, in-case the Questions were previously created and we introduce a new SubjectType / EncounterType, etc.. if we retrigger the AutoQuestionCreationAPI "/create-questions", would it work as expected and create Question only for the new one and not try and recreate all of them.? 
Ideally, Question creation should be an Idempotent operation.

**Comment by ombhardwajj** [comment]
Hey @himeshr I've addressed all the concerns including duplication of questions. Please review the code 🙇

**Comment by himeshr** [comment]
_"createQuestionForTable( schema => !public is not used)"_ => This one is pending, but going ahead with merge as its a small inconsequential thing.
